drivers/sensor/hx711/hx711.c

```diff
@@ -0,0 +1,562 @@
+ *        Formula is :
+ *          weight = slope * (reading - offset)
+ *          Only val1 is used.
+ *
+ * @return void
+ *
+ */
+static void hx711_attr_set_offset(struct hx711_data *data,
+				  const struct sensor_value *offset)
+{
+	data->offset = offset->val1;
+}
+
+/**
+ * @brief Set HX711 offset.
+ *
+ * @param dev Pointer to the hx711 device structure
+ * @param slope sensor_value struct. Slope used to calculate the real weight.
+ *        Formula is :
+ *          weight = slope * (reading - offset)
+ *
+ * @return void
+ *
+ */
+static void hx711_attr_set_slope(struct hx711_data *data,
+				 const struct sensor_value *slope)
+{
+	data->slope.val1 = slope->val1;
+	data->slope.val2 = slope->val2;
+}
+
+/**
+ * @brief Set HX711 attributes.
+ *
+ * @param dev Pointer to the hx711 device structure
+ * @param chan Channel to set.
+ *             Supported channels :
+ *               SENSOR_CHAN_WEIGHT
+ *               SENSOR_CHAN_ALL
+ * @param attr Attribute to change.
+ *             Supported attributes :
+ *               SENSOR_ATTR_SAMPLING_FREQUENCY
+ *               SENSOR_ATTR_OFFSET
+ *               SENSOR_ATTR_SLOPE
+ *               SENSOR_ATTR_GAIN
+ * @param val   Value to set.
+ * @retval  -ENOTSUP if an invalid attribute is given
+ *
+ */
+static int hx711_attr_set(const struct device *dev,
+			  enum sensor_channel chan,
+			  enum sensor_attribute attr,
+			  const struct sensor_value *val)
+{
+	int ret = 0;
+	struct hx711_data *data = dev->data;
+
+	switch (attr) {
+	case SENSOR_ATTR_SAMPLING_FREQUENCY:
+		ret = hx711_attr_set_rate(data, val);
+		if (ret == 0) {
+			LOG_DBG("Attribute RATE set to %d\n", data->rate);
+		}
+		return ret;
+	case SENSOR_ATTR_OFFSET:
+		hx711_attr_set_offset(data, val);
+		LOG_DBG("Attribute OFFSET set to %d\n", data->offset);
+		return ret;
+	case SENSOR_ATTR_SLOPE:
+		hx711_attr_set_slope(data, val);
+		LOG_DBG("Attribute SLOPE set to %d.%d\n",
+			data->slope.val1,
+			data->slope.val2);
+		return ret;
+	case SENSOR_ATTR_GAIN:
+		if (hx711_attr_set_gain(dev, val) == 0) {
+			LOG_DBG("Attribute GAIN set to %d\n", data->gain);
+		}
+		return ret;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+/**
+ * @brief Get HX711 reading.
+ *
+ * @param dev Pointer to the hx711 device structure
+ * @param chan Channel to set.
+ *             Supported channels :
+ *               SENSOR_CHAN_WEIGHT
+ *               SENSOR_CHAN_ALL
+ *
+ * @param val  Value to write weight value to.
+ *        Formula is :
+ *          weight = slope * (reading - offset)
+ *
+ * @retval 0 on success
+ * @retval  -ENOTSUP if an invalid channel is given
+ *
+ */
+static int hx711_channel_get(const struct device *dev,
+			     enum sensor_channel chan,
+			     struct sensor_value *val)
+{
+	struct hx711_data *data = dev->data;
+
+	switch (chan) {
+	case SENSOR_CHAN_WEIGHT:
+	case SENSOR_CHAN_ALL:
+		val->val1 = sensor_value_to_double(&data->slope)
```

Comment:
I would expect fractional part of the result to be available in `val2`, so something like this instead:
```c
case SENSOR_CHAN_ALL: {
	double result = sensor_value_to_double(&data->slope) *
			(data->reading - data->offset);
	val->val1 = (int32_t)result;
	val->val2 = (result - (int32_t)result) * 1000000;
	return 0;
}
```

Reply:
Absolutely Right.

Reply:
@nobodyguy, 
The only reason I opted then to only use **val1** was code simplification.

After thinking about it for awhile I believed it not to be _necessary_, as it would involve providing the **_target_** argument to _**avia_hx711_calibrate**_ as a **float**, which in turn would use a bit more **floating point arithmetic** and **sensor_value_to_double** calls.

Also currently **CONFIG_HX711_CALIBRATION_WEIGHT** is defined in KConfig.
Last time I checked floats were not supported so I decided to not do fractionals in  order to be able to provide configuration without hardcoding.

we could of course do
**CONFIG_HX711_CALIBRATION_WEIGHT_INTEGER_PART** and **CONFIG_HX711_CALIBRATION_WEIGHT_FRACTIONAL_PART**  but seems quite ugly  imho.

In a nutshell : if someone wants to use _fractional_ **Kg**, he should measure in **gr**  and divide :)

It is surely not unfeasible, and there are also the  **hx711_attr_set_slope** and **hx711_attr_set_offset** functions available.

What do you think about this?

Reply:
Fixed.